mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java

```diff
@@ -534,4 +537,370 @@ public void accessibleMessagesShouldFilterOutMessageIdsWhenNotBelongingToTheUser
+                    .forUser(MailboxFixture.OTHER_USER)
+                    .rights(MailboxACL.Rfc4314Rights.allExcept(MailboxACL.Right.DeleteMessages))
+                    .asAddition()),
+            aliceSession);
+        MessageId messageId = testingData.persist(aliceMailbox1.getMailboxId(), messageUid1, FLAGS, aliceSession);
+
+        //When
+        assertThatThrownBy(() ->
+            messageIdManager.setInMailboxes(messageId,
+                ImmutableList.of(bobMailbox1.getMailboxId()),
+                bobSession))
+            .isInstanceOf(MailboxNotFoundException.class);
+
+        //Then
+        List<MessageResult> messages = messageIdManager.getMessages(
+            ImmutableList.of(messageId),
+            FetchGroupImpl.MINIMAL,
+            bobSession);
+        assertThat(messages)
+            .extracting(MessageResult::getMessageId)
+            .containsOnly(messageId);
+        assertThat(messages)
+            .extracting(MessageResult::getMailboxId)
+            .containsOnly(aliceMailbox1.getMailboxId());
+    }
+
+    @Test
+    public void setInMailboxShouldAllowDistinctMailboxSetForShareeAndOwner() throws Exception {
+        //Given
+        testingData.setACL(aliceMailbox2.getMailboxId(),
+            MailboxACL.EMPTY.apply(
+                MailboxACL.command()
+                    .forUser(MailboxFixture.OTHER_USER)
+                    .rights(MailboxACL.FULL_RIGHTS)
+                    .asAddition()),
+            aliceSession);
+        MessageId messageId = testingData.persist(aliceMailbox1.getMailboxId(), messageUid1, FLAGS, aliceSession);
+        messageIdManager.setInMailboxes(messageId, ImmutableList.of(aliceMailbox1.getMailboxId(), aliceMailbox2.getMailboxId()), aliceSession);
+
+        //When
+        messageIdManager.setInMailboxes(messageId,
+            ImmutableList.of(bobMailbox1.getMailboxId(), aliceMailbox2.getMailboxId()),
+            bobSession);
+
+        //Then
+        List<MessageResult> messagesForSharee = messageIdManager.getMessages(
+            ImmutableList.of(messageId),
+            FetchGroupImpl.MINIMAL,
+            bobSession);
+        assertThat(messagesForSharee)
+            .extracting(MessageResult::getMessageId)
+            .containsOnly(messageId, messageId);
+        assertThat(messagesForSharee)
+            .extracting(MessageResult::getMailboxId)
+            .containsOnly(aliceMailbox2.getMailboxId(), bobMailbox1.getMailboxId());
+
+        List<MessageResult> messagesForOwner = messageIdManager.getMessages(
+            ImmutableList.of(messageId),
+            FetchGroupImpl.MINIMAL,
+            aliceSession);
+        assertThat(messagesForOwner)
+            .extracting(MessageResult::getMessageId)
+            .containsOnly(messageId, messageId);
+        assertThat(messagesForOwner)
+            .extracting(MessageResult::getMailboxId)
+            .containsOnly(aliceMailbox1.getMailboxId(), aliceMailbox2.getMailboxId());
```

Comment:
I don't see what is distinct

Reply:
Because alice copied from alice1 to alice2 so that bob view and alice view are not a subset of one another.

Alice1 != Bob1 by the way...

Reply:
yes sorry, that's ok, I thought I read it several times though.